### ISSUE TITLE
propTypes cycles cause crashes

### ISSUE THREAD

**Issue by ZxBing0066** [reactjs/react-docgen#640]
If there is a self-reference in propTypes define, docgen will throw an error: `RangeError: Maximum call stack size exceeded`. related to #249.

For reproduction, copy the code below to the playground: https://reactcommunity.org/react-docgen/

```jsx
import React, { Component } from 'react';
import PropTypes from 'prop-types';

/**
 * General component description.
 */
class MyComponent extends Component {
  render() {
    // ...
  }
}

MyComponent.propTypes = {
  /**
   * Description of prop "foo".
   */
  foo: PropTypes.number,
  /**
   * Description of prop "bar" (a custom validation function).
   */
  bar: function(props, propName, componentName) {
    // ...
  },
  baz: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string
  ]).isRequired,
  child: PropTypes.shape(MyComponent.propTypes)
};

MyComponent.defaultProps = {
  foo: 42,
  bar: 21
};

export default MyComponent;
```

**Comment by xunrongl** [comment]
running into the same issues with Flow, is there any plan to fix this soon? thanks!

**Comment by danez** [comment]
@xunrongl Can you please share an example with flow?

**Comment by xunrongl** [comment]
@danez sorry for delayed response. It is a little bit tricky to repro as it has quite complicated flow types, but let me try to provide you with an example

```
// AbstractComponentTypes.flow.js
export type ComponentRender<+T> = T | NestedComponentRender<T>
export type type NestedComponentRender<+T> =
  | React$Element<ClassComponentRender<T>>
  | React$Element<FunctionComponentRender<T>>;

export type NullableElement<+T> = ComponentRender<React$Element<T> | null> | null;
...

// FooComponent.flow.js
import typeof BarComponent from 'BarComponent.react';
import type {NullableElement} from 'AbstractComponentTypes.flow';
type A = NullableElement<BarComponent>;
export type Props = $ReadOnly<{
  propA: A,
  ...
}>;

// FooComponent.react.js
import type {Props} from 'FooComponent.flow';
...
```

and below is the error message
```
node_modules/react-docgen/dist/utils/printValue.js:49
    const src = getSrcFromAst(path);
                ^

RangeError: Maximum call stack size exceeded
    at printValue (third-party/node_modules/react-docgen/dist/utils/printValue.js:49:17)
    at Object.handleUnionTypeAnnotation [as UnionTypeAnnotation] (third-party/node_modules/react-docgen/dist/utils/getFlowType.js:174:39)
    at getFlowTypeWithResolvedTypes (third-party/node_modules/react-docgen/dist/utils/getFlowType.js:270:37)
    at Object.handleGenericTypeAnnotation [as GenericTypeAnnotation] (third-party/node_modules/react-docgen/dist/utils/getFlowType.js:109:16)
    at getFlowTypeWithResolvedTypes (third-party/node_modules/react-docgen/dist/utils/getFlowType.js:270:37)
    at NodePath.<anonymous> (third-party/node_modules/react-docgen/dist/utils/getFlowType.js:118:45)
    at NodePath.<anonymous> (third-party/node_modules/ast-types/lib/path.js:94:34)
    at NodePath.each (third-party/node_modules/ast-types/lib/path.js:87:26)
    at NodePath.map (third-party/node_modules/ast-types/lib/path.js:93:14)
    at Object.handleGenericTypeAnnotation [as GenericTypeAnnotation] (third-party/node_modules/react-docgen/dist/utils/getFlowType.js:118:30)
```

Let me know if you need more information. thanks

**Comment by danez** [comment]
The initial issue is resolved in v6.0.0-beta.5.

The flow issue I wasn't able to reproduce. Looking at the changelog I see that it was using version 5 of react-docgen, so maybe see if it is fixed in version 6. If not please open a new issue with a minimal reproducible example. Thanks.
